peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+
+The ``TypeExpr`` special form is not a ``type`` at runtime. It is
+however an ``object``:
+
+::
+
+   >>> isinstance(TypeExpr, type)
+   False
+
+   >>> isinstance(TypeExpr, object)
+   True
+
+The ``TypeExpr`` special form cannot be used as the ``type`` argument to
+``isinstance``:
+
+::
+
+   >>> isinstance(str, TypeExpr)
+   TypeError: typing.TypeExpr cannot be used with isinstance()
+
+   >>> isinstance(str, TypeExpr[str])
+   TypeError: isinstance() argument 2 cannot be a parameterized generic
+
+The ``TypeExpr`` special form cannot be used as any argument to
+``issubclass``:
+
+::
+
+   >>> issubclass(TypeExpr, object)
+   TypeError: issubclass() arg 1 must be a class
+
+   >>> issubclass(object, TypeExpr)
+   TypeError: typing.TypeExpr cannot be used with issubclass()
+
+
+Affected signatures in the standard library
+-------------------------------------------
+
+Changed signatures
+''''''''''''''''''
+
+The following signatures related to type expressions introduce
+``TypeExpr`` where previously ``object`` existed:
+
+-  ``typing.cast(typ: TypeExpr[T], val: object) -> T``
+-  ``typing.assert_type(val: object, typ: TypeExpr, /) -> None``
+
+
+Unchanged signatures
+''''''''''''''''''''
+
+The following signatures related to annotation expressions continue to
+use ``object`` and remain unchanged:
+
+-  ``typing.get_origin(tp: object) -> object``
+-  ``typing.get_args(tp: object) -> tuple[object, ...]``
+
+The following signatures related to class objects continue to use
+``type`` and remain unchanged:
+
+-  ``builtins.isinstance(object: object, classinfo: type | tuple[type, ...]) -> bool``
+-  ``builtins.issubclass(class: type, classinfo: type | tuple[type, ...]) -> bool``
+-  ::
+
+        @overload
+        builtins.type(object: object) -> type
+        @overload
+        builtins.type(
+            name: str,
+            bases: tuple[type, ...],
+            dict: dict[str, object],
+            **kwds
+        ) -> type
+
+``typing.get_type_hints(..., include_extras=False)`` nearly returns only type
+expressions in Python 3.12, stripping out most type qualifiers
+(``Required, NotRequired, ReadOnly, Annotated``) but currently preserves a
+few type qualifiers which are only allowed in annotation expressions
+(``ClassVar, Final, InitVar, Unpack``). It may be desirable to alter the
+behavior of this function in the future to also strip out those
+qualifiers and actually return type expressions, although this PEP does
+not propose those changes now:
+
+-  ::
+
+        @overload
+        typing.get_type_hints(obj: object, ..., include_extras: Literal[False]) 
+            -> dict[str, object]  # almost returns type expressions, but not quite
+        @overload
+        typing.get_type_hints(obj: object, ..., include_extras: Literal[True]) 
+            -> dict[str, object]  # returns annotation expressions
+
+
+Backwards Compatibility
+=======================
+
+No backward incompatible changes are made by this PEP.
+
+
+How to Teach This
```

Comment:
A question that came up at the Typing Summit that isn't explicitly addressed here: How is the *body* of a function that takes a `TypeExpr` parameter type checked? The answer is that it isn't type checked any more than if you had taken an `object` parameter, since static type checkers don't know about the runtime representation of different TypeExprs. I think it's worth spelling that out explicitly in the PEP.